### PR TITLE
[WIP] Fix brace matching for outer braces

### LINKED ISSUE #3794: Outer parens are not highlighted

### PR DESCRIPTION
**Pull request by cartermp**
Fixes #3794 

Not entirely sure if this is the _right_ fix, but it certainly is _a_ fix.  It's really just the `range.Contains()` call that fixes this, though.

Needs test updates

### CODE REVIEW

**Comment by cartermp** [comment]
So Roslyn is actually calling our `IBraceMatcher` implementation four times:

four-times

Reverting the check to just `range.Contains(position)` will return false for `isPositionInrange` in the _first_ call, whereas with this PR it will will return true.  I have no idea why that is, but I think it's strange that we're getting called so often.  Will have to debug Roslyn to see what's going on.

File: vsintegration/src/FSharp.Editor/Formatting/BraceMatchingService.fs [hunk 1]
```diff
@@ -14,28 +13,33 @@ type internal FSharpBraceMatchingService
...
                 | None -> false
                 | Some range ->
-                    let length = position - range.Start
-                    length >= 0 && length <= range.Length
+                    // We need to see if the caret is contained within or on the outside of the span
+                    // that we get back from the language service.
+                    //
+                    // Ex: let x = ((12))^
+                    //
+                    // The caret can be on the outside of the last paren, but this is actually 1 position
```
Comment (cartermp) [inline vsintegration/src/FSharp.Editor/Formatting/BraceMatchingService.fs]: this doesn't appear to be true at all

**Comment by vasily-kirichenko** [comment]
Maybe a test or two should be added to https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/tests/unittests/BraceMatchingServiceTests.fs?

File: vsintegration/src/FSharp.Editor/Formatting/BraceMatchingService.fs [hunk 2]
```diff
@@ -14,28 +13,33 @@ type internal FSharpBraceMatchingService
...
+                    // further to the right than the end of the span that we get back.
+                    range.Contains(caretPosition) || range.End = caretPosition + 1
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Formatting/BraceMatchingService.fs]: Shouldn't this be `range.End = caretPosition`?

Comment (cartermp) [reply vsintegration/src/FSharp.Editor/Formatting/BraceMatchingService.fs]: That's what I originally thought, but it it results in the same behavior as #3794 (and the current implementation).

Changing to just `range.Contains(caretPosition)` fixes #3794 and also results the same behavior as the current implementation for #2092 (all cases except for `let (| A9 |)` and the caret being _inside_ a brace pair result in highlighting).

Unfortunately, debugging this is a real mess.  I think it has something to do with debugging CEs, but `isPositionInRange` is called twice as often as I would expect when debugging.

**Comment by cartermp** [comment]
Actually @saul back to this comment: https://github.com/Microsoft/visualfsharp/issues/2092#issuecomment-326741943

You fixed brace matching such that (given `^` is the caret position):

```fsharp
 let x= ( 12 ^)
```
Will match those braces.  I don't think that's the correct thing to do.  Given this:

```fsharp
let x = ((12)^)
```
It's not clear which brace should be highlighted.  Roslyn throws out the outer pair and so the inner one is matched, but I'm not convinced that we should be giving them both pairs.  Thoughts?

**Comment by saul** [comment]
I'm not entirely sure - does it really matter? I was just trying to fix the bug raised in #2092 which I agreed with. My change was to fix most of those issues, and to also fix the logic for auto-deindent (which I believe this PR breaks, looking at the tests)

**Comment by cartermp** [comment]
Not sure if it matters, but given that ambiguity I'm leaning towards not counting that case.

What is a bit baffling is that all the test cases in `BraceMatchingBothSides_Bug2092` fail if the caret is at the very end of the last brace, but when I run the code outside the context of a test they highlight just fine.  Debugging the text shows that `span.Contains(caretPosition)` clearly fails, but that doesn't appear to matter when the code is being run in the editor.  This also breaks auto-deindent as well, so there's something funky going on.

**Comment by cartermp** [comment]
When debugging the editor, we're getting called twice:

twice

`caretPosition` is different the second time.  No clue why that is.

**Comment by realvictorprm** [comment]
@cartermp is this still valid?

**Comment by cartermp** [comment]
Yes. The current brace matching implementation doesn't work because of some weirdness with Roslyn calling us twice and passing different values.

**Comment by TIHan** [comment]
@cartermp There is a reason why it is passing two different values. http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/BraceMatching/BraceHighlightingViewTaggerProvider.cs,77

This is intended behavior. To summarize, here is an example (C#):
```csharp
()^()
```
It will check the right side for a starting brace; if there is, it will match. (position)
Then, it will check the left side for an ending brace; if there is, it will match. (position - 1)
tagme

I also want to note that `IBraceMatcher` is an internal API in Roslyn. In the future, we either implement our own version of all of this using [IViewTaggerProvider](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.tagging.iviewtaggerprovider?view=visualstudiosdk-2017) or convince Roslyn to make this public.

**Comment by realvictorprm** [comment]
@TIHan if it's about reimplementing taggers I can help out here. Thanks to code lens I know too much about them.

**Comment by cartermp** [comment]
With regards to the API being internal - that's the case for, unfortunately, too many of the things we implement or override.  The best thing is to help convince folks to make all the workspace APIs we deal with public so that we can get rid of the internalsvisibleto hack.

**Comment by realvictorprm** [comment]
Can the community help with convincing them? E.g. writing nice blog posts 😝?

**Comment by cartermp** [comment]
Not really, no. It's a big cost for a small team with a huge feature area to open them up right now. It affects more than just F#, though, so I'm hopeful we can get to it in 2018.

**Comment by cartermp** [comment]
Closing this, as there is likely to be more advanced logic required to fix this problem correctly.

**Pull request closed without merging**